### PR TITLE
Add support for class based plugins

### PR DESCRIPTION
**Pull request by SerayaEryn**
I started looking into a possible implementation of class based plugins (#1655).

There are decorators for each http method: `@Get`, `@Post`, `@Put`, `@Delete`, `@Put`, `@Head`, `@Patch` and `@Options`. In addition to that there is a `@All` decorator that adds the handler for all http methods. They accept the parameters like the methods on the fastify instance: `@Method(path, [options])`
Furthermore there is a `@Hook(hook)` decorator that allows to add hooks.

These decorators for the http methods and the hook decorator seem to work well.

I also have build decorators, that allow to decorate the fastify instance, request and reply.
However I don't know whether this solution is nice.
It is not possible to check whether the fastify instance, request or reply is already decorated.
And since a function can't access the `this` from the plugin and the request (or reply) the decorated method needs to return another function.
```ts
class MyPlugin {
  @DecorateRequest('utility')
  addUtility () {
    return function () {
      // something very useful
    }
  }
}
```

Maybe the following approach might be better:

```ts
class MyPlugin {
  @SupplyWithInstance // or something similar
  decorate (instance) {
    if(!instance.hasRequestDecorator('utility')) {
      instance.decorateRequest('utility', function () {
        // something very useful
      })
    }
  }
}
```

The documentation for the class based plugins is not finished yet.

I think this feature should be marked as experimental as long as the typescript feature is experimental and the proposal is not finalized. And maybe there should be an option that needs to be enabled?

I also noticed that it might be nice to add some kind of `@Plugin(metadata)` decorator to `fastify-plugin`, that allows to add metadata for a plugin.

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: lib/pluginDecorators.js [hunk 1]
```diff
@@ -0,0 +1,70 @@
+'use strict'
+
+const { kPluginMetadata } = require('./symbols')
+require('reflect-metadata')
```
Comment (mcollina) [inline lib/pluginDecorators.js]: I'm -1 on adding any dependency that modifies global state. Can we do this without this library or with a work-around?

File: test/classBasedPlugins/tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,11 @@
+{
+  "compilerOptions": {
+    "target": "es6",
+    "module": "commonjs",
+    "strict": false,
+    "experimentalDecorators": true
```
Comment (mcollina) [inline test/classBasedPlugins/tsconfig.json]: I _think_ I've heard they are going to change this in the next version of TS. When will decorators be out of experimental? Have you got any info?

File: docs/ClassBasedPlugins.md [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+**Note**: It is required to enable the `experimentalDecorators` compiler option to be able to use the typescript decorators.
+
+```ts
+class MyPlugin {
+  constructor (options) {
+    // ...
+  }
+
+  @Get('/')
+  async route (request, reply) {
```
Comment (mcollina) [inline docs/ClassBasedPlugins.md]: I think we should expose the current fastify plugin instance via `this.instance`  or `this.app`. Some decorators will be available there.

File: docs/ClassBasedPlugins.md [hunk 4]
```diff
@@ -0,0 +1,131 @@
+
+# Class based plugins
+
+> Experimental
+
+**Note**: It is required to enable the `experimentalDecorators` compiler option to be able to use the typescript decorators.
+
+```ts
+class MyPlugin {
```
Comment (mcollina) [inline docs/ClassBasedPlugins.md]: How can we add a prefix for this plugin?

File: docs/ClassBasedPlugins.md [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+}
+```
+
+## Decoration of request, reply or instance
+
+```ts
+import { DecorateInstance } from 'fastify'
+
+class MyPlugin {
+  @DecorateInstance()
```
Comment (mcollina) [inline docs/ClassBasedPlugins.md]: We would need to add a decorator on the class itself to enable/disable encapsulation, i.e. setting https://github.com/fastify/fastify-plugin/blob/966248ce75eebe58727aaa5c676d410b4641d236/index.js#L16

**Comment by Ethan-Arrowood** [comment]
Are we planning on landing this in v2?

**Comment by mcollina** [comment]
I would prefer typescript decorators to exit experimental status first.
Likely v3.

Il giorno mar 28 mag 2019 alle 19:31 Ethan Arrowood <
notifications@github.com> ha scritto:

File: package.json [hunk 6]
```diff
@@ -11,9 +11,10 @@
     "unit": "tap -J test/*.test.js test/*/*.test.js",
     "unit:report": "tap -J test/*.test.js test/*/*.test.js --cov --coverage-report=html --coverage-report=cobertura | tee out.tap",
     "unit:junit": "tap-mocha-reporter xunit < out.tap > test/junit-testresults.xml",
+    "unit:compile": "tsc --project ./test/classBasedPlugins/tsconfig.json",
     "typescript": "tsc --project ./test/types/tsconfig.json",
-    "test:report": "npm run lint && npm run unit:report && npm run typescript",
-    "test": "npm run lint && npm run unit && npm run typescript",
+    "test:report": "npm run lint && npm run unit:compile && npm run unit:report && npm run typescript",
```
Comment (Ethan-Arrowood) [inline package.json]: I'm concerned with how these .ts tests will affect the coverage report. Can you verify the current test coverage reporter is accurately reporting that the new API code is being covered by these tests?

**Comment by Ethan-Arrowood** [comment]
Per @mcollina confirmation, please rebase onto `next` branch, and similarly, change the PR base is to `next` instead of `master`. I recently did this for the typescript refactor PR so let me know if you have issues/questions 😁👍

File: lib/pluginDecorators.js (hunk 1 above)
Comment (SerayaEryn) [reply lib/pluginDecorators.js]: We could store it at the plugin using a symbol.

File: docs/ClassBasedPlugins.md (hunk 5 above)
Comment (SerayaEryn) [reply docs/ClassBasedPlugins.md]: I agree. Shall we add this decorator in `fastify-plugin`? It could provide the other features of `fastify-plugin` as a decorator, too.

File: package.json (hunk 6 above)
Comment (SerayaEryn) [reply package.json]: The coverage is being reported correctly if the test have been compiled.
However I forgot compile the typescript tests in `coverage` and `coveralls` tasks. I'll change that.

File: docs/ClassBasedPlugins.md (hunk 4 above)
Comment (SerayaEryn) [reply docs/ClassBasedPlugins.md]: With the prefix option, that is being used for the normal plugins, too:
```js
fastify.register(new MyPlugin(), { prefix: ''/foo' })
```

File: docs/ClassBasedPlugins.md (hunk 3 above)
Comment (SerayaEryn) [reply docs/ClassBasedPlugins.md]: I'll do that.

File: test/classBasedPlugins/tsconfig.json (hunk 2 above)
Comment (SerayaEryn) [reply test/classBasedPlugins/tsconfig.json]: No, I got no information on this.
However I think we will know in two days: https://github.com/microsoft/TypeScript/issues/30555

**Comment by SerayaEryn** [comment]
Ok, I'll rebase the branch and change the PR.

File: test/classBasedPlugins/tsconfig.json (hunk 2 above)
Comment (SerayaEryn) [reply test/classBasedPlugins/tsconfig.json]: Decorators are still experimental in typescript 3.5.x.

File: docs/ClassBasedPlugins.md (hunk 3 above)
Comment (SerayaEryn) [reply docs/ClassBasedPlugins.md]: The current fastify instance is now being exposed.

File: lib/abstractPlugin.js [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+const { kSetInstance } = require('./symbols.js')
+
+module.exports = class AbstractPlugin {
+  constructor () {
+    this.instance = null
+  }
+
+  [kSetInstance] (instance) {
+    this.instance = instance
+  }
```
Comment (mcollina) [inline lib/abstractPlugin.js]: I don’t think this class is needed at all. We can just set the instance manually on the object, without using an accessor.

File: test/classBasedPlugins/class-based-plugin.test.ts [hunk 8]
```diff
@@ -218,10 +217,10 @@ test('Should decorate reply', (t) => {
 test('Should expose fastify instance to plugin', (t) => {
   t.plan(5)
   
-  class TestPlugin extends AbstractPlugin {
+  class TestPlugin {
     @Get('/')
     async handler (request, response) {
-      t.ok(this.instance)
+      t.ok(this['instance'])
```
Comment (mcollina) [inline test/classBasedPlugins/class-based-plugin.test.ts]: why this change?

File: fastify.js [hunk 9]
```diff
@@ -765,3 +781,15 @@ function buildRoutePrefix (instancePrefix, pluginPrefix) {
...
+module.exports.Delete = Delete
+module.exports.Head = Head
+module.exports.Put = Put
+module.exports.Options = Options
+module.exports.Patch = Patch
+module.exports.All = All
+module.exports.Hook = Hook
+module.exports.DecorateRequest = DecorateRequest
+module.exports.DecorateReply = DecorateReply
+module.exports.DecorateInstance = DecorateInstance
```
Comment (mcollina) [inline fastify.js]: Would it be possible to use these with JavaScript as well?

**Comment by mcollina** [comment]
I really like this proposal. I'm a bit scared about how this would affect our LTS support, considering that  TS decorators are still experimental.

Should we land them as experimental here, maybe logging a warning?

File: fastify.js (hunk 9 above)
Comment (SerayaEryn) [reply fastify.js]: The decorators can be used with babel.
Using them without typescript or babel should be possible, but it is not very practical.

File: test/classBasedPlugins/class-based-plugin.test.ts (hunk 8 above)
Comment (SerayaEryn) [reply test/classBasedPlugins/class-based-plugin.test.ts]: That wasn't planed to stay. I fixed it.

**Comment by SerayaEryn** [comment]
I agree that they should be experimental and a warnung is a good idea.
I'll add a warning.

File: docs/ClassBasedPlugins.md (hunk 5 above)
Comment (SerayaEryn) [reply docs/ClassBasedPlugins.md]: I opened a PR (fastify/fastify-plugin#73) in the fastify-plugin repository that adds such a decorator.

**Comment by Ethan-Arrowood** [comment]
If this is to land in `next` I'd like it to be experimental until decorators are standard. 

Additionally, when #1569 lands I believe some updates will need to be made with this proposal.

**Comment by SerayaEryn** [comment]
Just a note: I am waiting until #1569 is being merged before I add typings for the new Decorators that this PR adds.

**Comment by areese159** [comment]
I'm curious, how these would work with multiple instances of Fastify running in node? Right now we have a public server that hosts all of our authenticated routes to the public and internal server that handles our internal traffic from other services. With this implementation, it does not appear that this will be possible. Is that something that we could support or will decorators only support running a single instance of Fastify?

**Comment by SerayaEryn** [comment]
I rebased the branch since #1569 has been merged and added typings for the decorators.

**Comment by SerayaEryn** [comment]
The test failures seem to be related to `typescript@3.6.2`. With `typescript@3.5.3` all tests are passing.

**Comment by RafaelGSS** [comment]
Hi @SerayaEryn! Really excellent addition. 
But I have some doubts, based on this example retired of doc:

```ts
class MyPlugin {
  constructor (options) {
    // ...
  }
  @Get('/')
  async route (request, reply) {
    return 'hello world'
  }
}
```

- How I can add a `preHandler` for each route?
- The same as occurs after I do #1828

**Comment by SerayaEryn** [comment]
The decorators for the http methods have a second parameter that allows to add additional route options:

```js
class MyPlugin {
  constructor (options) {
    // ...
  }
  @Get('/', { preHandler: yourPreHandler })
  async route (request, reply) {
    return 'hello world'
  }
}
```

**Comment by mcollina** [comment]
This work is impressive.

CI is failing, would you mind taking a look?

**Comment by SerayaEryn** [comment]
The new `typescript` version `3.6.2` seems to cause the failing tests. I got the same failing tests with that version on my local machine. With the older `3.5.3` version they pass.

I take a closer look next week.

**Comment by StarpTech** [comment]
Great PR but I don't like that code is shipped to the core and only typescript is supported.

**Comment by StarpTech** [comment]
Can we provide them as a plugin?

**Comment by SerayaEryn** [comment]
> Can we provide them as a plugin?

I think that should be possible.

**Comment by mcollina** [comment]
If we can pack them as a plugin, it would be fantastic.

Il giorno mer 18 set 2019 alle ore 20:45 Denis Fäcke <
notifications@github.com> ha scritto:

**Comment by SerayaEryn** [comment]
I'll close this PR and will start working on a module as soon as i can.
